webform_civicrm_membership_extras.module

```diff
@@ -762,14 +762,94 @@ function _webform_civicrm_membership_extras_get_membership_types(&$form, &$form_
  * @param array $membershipNamesAndKeysMap
  */
 function _webform_civicrm_membership_extras_get_membership_dates(&$form, &$form_state, $membershipTypes, $membershipNamesAndKeysMap) {
+  $membership_dates_added = [];
+  $components = _webform_civicrm_membership_extras_get_components_by_part_of_form_key($form, '_membership_start_date');
+  foreach ($components as $key => $component) {
+    $start_date = 0;
+    foreach($form_state['input']['submitted'] as $fieldset) {
+      $day = $fieldset[$component['form_key']]['day'] ?? 0;
+      $month = $fieldset[$component['form_key']]['month'] ?? 0;
+      $year = $fieldset[$component['form_key']]['year'] ?? 0;
+      if ($day && $month && $year) {
+        $start_date = intval($year) . "-" . intval($month) . "-" . intval($day);
+      }
+    }
+
+    if (empty($start_date)) {
+      continue;
+    }
+    $membership_number = explode('_membership_start_date', $component['form_key'])[0];
+    $membership_type_key = $membershipNamesAndKeysMap[$membership_number];
+    $membership_dates_added[$membership_type_key]['start_date'] = $start_date;
+  }
+
+  $components = _webform_civicrm_membership_extras_get_components_by_part_of_form_key($form, '_membership_end_date');
+  foreach ($components as $key => $component) {
+    $end_date = 0;
+    foreach($form_state['input']['submitted'] as $fieldset) {
+      $day = $fieldset[$component['form_key']]['day'] ?? 0;
+      $month = $fieldset[$component['form_key']]['month'] ?? 0;
+      $year = $fieldset[$component['form_key']]['year'] ?? 0;
+      if ($day && $month && $year) {
+        $end_date = intval($year) . "-" . intval($month) . "-" . intval($day);
+      }
+    }
+
+    if (empty($end_date)) {
+      continue;
+    }
+    $membership_number = explode('_membership_end_date', $component['form_key'])[0];
+    $membership_type_key = $membershipNamesAndKeysMap[$membership_number];
+    $membership_dates_added[$membership_type_key]['end_date'] = $end_date;
+  }
+
+  $components = _webform_civicrm_membership_extras_get_components_by_part_of_form_key($form, '_membership_join_date');
```

Comment:
seems there is code repetition in this method : _webform_civicrm_membership_extras_get_membership_dates 

maybe you can extract the similar code into another function.

Reply:
I've refactored the code and created two more functions.